src/flash/nor/mspm0.c: Fix byte size issue and resolve write alignmen…

HenryNguyen01 opened the pull request:
…t concern

Resolved byte size issue with target_write_memory() API. Also updated probe to enforce the alignment values for start and end of writes. With the write alignment variable it is set to 4 because due to the fact we utilize only CMDDATA0 which is 32-bits (8-bytes).

Change-Id: Ib4fd963d22d8c9a2f0358dc96081cac3a2285c71

HNguyen00 committed:
src/flash/nor/mspm0.c: Fix byte size issue and resolve write alignment concern

Resolved byte size issue with target_write_memory() API. Also updated probe
to enforce the alignment values for start and end of writes. With the write
alignment variable it is set to 4 because due to the fact we utilize only
CMDDATA0 which is 32-bits (8-bytes).

Change-Id: Ib4fd963d22d8c9a2f0358dc96081cac3a2285c71
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>

nmenon commented on src/flash/nor/mspm0.c:
```diff
@@ -1026,7 +1020,7 @@ static int mspm0_write(struct flash_bank *bank, const unsigned char *buffer,
 
-		retval = target_write_memory(target, FCTL_REG_CMDDATA0, 1, num_bytes_to_write, buffer);
+		retval = target_write_memory(target, FCTL_REG_CMDDATA0, 4, num_bytes_to_write / 4, buffer);
 		if (retval)
```
is there a case where num_bytes_to_write is non-multiple of 4?

HenryNguyen01 replied:
Do you really want to write bytes? Check target_memory_write() doc.
To keep writing by 32-bit words as was in the previous code use <--------old comment from Tomas

In my opinion the previous implementation made more sense as the third parameter is the size and fourth would be the count? 

Attaching ss of the API 
![image](https://github.com/user-attachments/assets/971f3cee-ad6b-4bae-b98a-8ddedb094da0)

HenryNguyen01 replied:
![image](https://github.com/user-attachments/assets/9f2629e7-3e3b-47e1-a996-e5c0646c36a7)
Here's the documentation of the API online.
From what it's describing wouldn't it better actually to keep it as how it was done before?

HenryNguyen01 replied:
From what I gathered in the docs, the api will write x (count) amount of y (size) bytes to the location.

"Write 'count' items of 'size' bytes"

Wouldn't it be better in this case actually to keep it as is to account for the scenario where the amount being written or "num_bytes_to_write" is not a multiple of 4?

nmenon commented on src/flash/nor/mspm0.c:
```diff
@@ -1092,6 +1086,9 @@ static int mspm0_probe(struct flash_bank *bank)
 	}
 
+	bank->write_start_alignment = 4;
+	bank->write_end_alignment = 4;
```
is this why writes are going to be fine?

HenryNguyen01 replied:
According to the documentation and the implementation it should be 👀 

When setting the alignment value it forces the address to be a multiple of whatever it's set to and since we write 32-bits at a time it should be fine

HenryNguyen01 commented:
Realized the mistake in my PR comment but I meant to say 4-bytes not 8 😅

HNguyen00 committed:
src/flash/nor/mspm0.c: Pivot to target_write_memory to target_write_buffer

Updating the api utilized to write to CMDDATA0 for a more flexible/robust
approach.

Change-Id: I39934c5dfb1e6a04fb0f28f980a0ab631c3f9a5f
Signed-off-by: Henry Nguyen <h-nguyen8@ti.com>
